src/Common/FST.h

```diff
@@ -0,0 +1,171 @@
+        };
+        State() = default;
+        State(const State & state) = default;
+
+        UInt64 hash() const;
+
+        Arc * getArc(char label);
+        void addArc(char label, Output output, StatePtr target);
+
+        void clear()
+        {
+            id = 0;
+            state_index = 0;
+            flag = 0;
+
+            arcs.clear();
+        }
+
+        UInt64 serialize(WriteBuffer & write_buffer);
+
+        inline bool isFinal() const
+        {
+            return flag_values.is_final == 1;
+        }
+        inline void setFinal(bool value)
+        {
+            flag_values.is_final = value;
+        }
+        inline EncodingMethod getEncodingMethod() const
+        {
+            return flag_values.encoding_method;
+        }
+        inline void readFlag(ReadBuffer & read_buffer)
+        {
+            read_buffer.read(reinterpret_cast<char&>(flag));
+        }
+
+        UInt64 id = 0;
+        UInt64 state_index = 0;
+        std::unordered_map<char, Arc> arcs;
+    private:
+        struct FlagValues
+        {
+            unsigned int is_final : 1;
+            EncodingMethod encoding_method : 3;
+        };
+
+        union
+        {
+            FlagValues flag_values;
+            uint8_t flag = 0;
+        };
+    };
+
+    bool operator==(const State & state1, const State & state2);
+
+    inline constexpr size_t MAX_TERM_LENGTH = 256;
+
+    /// FSTBuilder is used to build Finite State Transducer by adding words incrementally.
+    /// Note that all the words have to be added in sorted order in order to achieve minimized result.
+    /// In the end, the caller should call build() to serialize minimized FST to WriteBuffer
+    class FSTBuilder
+    {
+    public:
+        FSTBuilder(WriteBuffer & write_buffer_);
+        StatePtr getMinimized(const State & s, bool & found);
+
+        void add(const std::string & word, Output output);
+        UInt64 build();
+
+        UInt64 state_count = 0;
+
+    private:
+        void minimizePreviousWordSuffix(Int64 down_to);
+        static size_t getCommonPrefix(const std::string & word1, const std::string & word2);
+
+        std::array<StatePtr, MAX_TERM_LENGTH + 1> temp_states;
+        std::string previous_word;
+        StatePtr initial_state;
+        std::unordered_map<UInt64, StatePtr> minimized_states;
+
+        UInt64 next_id = 1;
+
+        WriteBuffer & write_buffer;
+        UInt64 previous_written_bytes = 0;
+        UInt64 previous_state_index = 0;
+    };
+
+    //FiniteStateTransducer is constructed by using minimized FST blob(which is loaded from index storage)
+    // It is used to retrieve output by given term
+    class FiniteStateTransducer
+    {
+    public:
+        FiniteStateTransducer() = default;
+        FiniteStateTransducer(std::vector<UInt8> data_);
+        std::pair<UInt64, bool> getOutput(const String & term);
+        void clear();
+        std::vector<UInt8> & getData() { return data; }
```

Comment:
getData() is only used when reading data into FST from files. We could introduce `read()` method, but it would make FST more dependent to other classes. So let's keep it simple, IMHO.